Oeuf

Dune-jr opened the pull request.

oy commented on src/game/client/components/maplayers.cpp:
```diff
@@ -81,12 +81,127 @@ void CMapLayers::OnInit()
+
+void CMapLayers::OnShutdown()
+{
+	mem_free(m_aEggTiles);
```
check if m_aEggTiles is nonzero

oy commented on src/game/client/components/maplayers.h:
```diff
@@ -19,6 +19,10 @@ class CMapLayers : public CComponent
 	array<CEnvPoint> m_lEnvPoints;
 	array<CEnvPoint> m_lEnvPointsMenu;
 
+	CTile* m_aEggTiles;
```
rename to m_pEggTiles

Dune-jr commented:
Fixed.

Dune-jr commented:
As suggested @LordSk, this might be best kept for Easter. After all, it fits the Easter theme better. Would have to implement [Computus](https://en.wikipedia.org/wiki/Computus).

oy commented:
Yeah, that sounds good :)

Dune-jr commented:
Implemented Computus for the Gregorian Calendar. `time_iseasterday()` now triggers for a range of 4 days (Good Friday, Easter Saturday, Easter Sunday, Easter Monday).
Different subcultures use different days for Easter celebration (any of those 4 afaik).

The Gregorian Calendar is used by the Catholic and Protestant Churches, but not by the Eastern Churches. This means that Easter typically comes a week late for Eastern Europe/Russia (sometimes much more). I see three options:

1. Trigger for Gregorian and Julian Easter week-ends (with a non-Easter period in between)
1. Trigger from Gregorian Easter to Julian Easter only when the two are close (~8 days) and not 30 days apart. This is true in 2019, 2020, false in 2021.
1. Trigger only for Gregorian calendars

Option 1. seems meh to me because of the hole in between; it is probably going to confuse a lot of westerners when the second event occurs.
Option 2. seems bad to me because it is inconsistent (and kind of long).
Option 3 seems like the best bet to me. There is a significant amount of Catholics in Russia and I believe they celebrate there as well so I think Russians will at least understand what's going on.


## Testing
The C code was taken from Wikipedia and I tested it with Linux's ncal:
```cpp

int main()
{
	srand(time(NULL));
	int Y = (rand()%300)+1800;

	int a = Y % 19;
	int b = Y / 100;
	int c = Y % 100;
	int d = b / 4;
	int e = b % 4;
	int f = (b + 8) / 25;
	int g = (b - f + 1) / 3;
	int h = (19 * a + b - d - g + 15) % 30;
	int i = c / 4;
	int k = c % 4;
	int L = (32 + 2 * e + 2 * i - h - k) % 7;
	int m = (a + 11 * h + 22 * L) / 451;
	int month = (h + L - 7 * m + 114) / 31;
	int day = ((h + L - 7 * m + 114) % 31) + 1;

	fprintf(stderr, "%d/%d\n", day, month);
	fprintf(stdout, "%d", Y);
	return Y;
}
```
```
~/tmp/tw $ ncal -e `./a.out`
26/3
26 mars 1826
~/tmp/tw $ ncal -e `./a.out`
4/4
 4 avril 2094
~/tmp/tw $ ncal -e `./a.out`
10/4
10 avril 2039
~/tmp/tw $ ncal -e `./a.out`
14/4
14 avril 1963
~/tmp/tw $ ncal -e `./a.out`
4/4
 4 avril 1926
~/tmp/tw $ ncal -e `./a.out`
29/3
29 mars 1964
~/tmp/tw $ ncal -e `./a.out`
29/3
29 mars 1964
~/tmp/tw $ ncal -e `./a.out`
4/4
 4 avril 2083
~/tmp/tw $ ncal -e `./a.out`
13/4
13 avril 1873
~/tmp/tw $ ncal -e `./a.out`
```

### Now testing that the `time_iseasterday` returns the right value using:
```cpp

int main(int argc, char** argv)
{
	int X;
	sscanf(argv[1], "%d", &X);
	printf("time_iseasterday=%d\n", time_iseasterday(X));
	return 0;
}

int time_iseasterday(int X)
{
	time_t time_data_now, time_data;
	struct tm *time_info;

	// time(&time_data_now);
	time_data_now = X;
	time_info = localtime(&time_data_now);
```
This gives:
```
$ ./a.out `date --date="Apr 22 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=1
$ ./a.out `date --date="Apr 22 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=1
$ ./a.out `date --date="Apr 21 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=1
$ ./a.out `date --date="Apr 20 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=1
$ ./a.out `date --date="Apr 19 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=1
$ ./a.out `date --date="Apr 18 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=0
$ ./a.out `date --date="Apr 17 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=0
$ ./a.out `date --date="Apr 23 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=0
$ ./a.out `date --date="May 23 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=0
$ ./a.out `date --date="May 21 09:00:00 UTC+0200 2019" +%s`
time_iseasterday=0
$ ./a.out `date --date="Apr 12 09:00:00 UTC+0200 2020" +%s`
time_iseasterday=1
$ ncal -e 2020
12 avril 2020
```

Found and corrected two bugs with the testing, looks good now

Dune-jr commented:
https://github.com/teeworlds/teeworlds/commit/ac463bf1d74d06cb2a29dda14031dbaf796977ea
